Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+```
+
+And I hope that with this you can better understand how parser combinators work. Now get to hacking your languages!
+
+The complete code for this part can be found [here](https://gist.github.com/heitor-lassarote/ea54d05aca2956efa29b47ebbf048cbd).
+
+## Bonus: QuasiQuotation
+
+<!--
+* show how to use QQ to parse in compile-time
+-->
+
+This section is better understood if you have some Template Haskell (TH) knowledge. If you haven't learned about it, then I **strongly** encourage you to check out our [Introduction to Template Haskell](https://serokell.io/blog/introduction-to-template-haskell) before trying out this section.
+
+Note that this section is not necessary for our knowledge of parser combinators, and it's purely a bonus section on how to do more with our S-expressions.
+
+For this part, you will need the [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package.
+
+Long story short, Template Haskell is a mechanism that allows GHC to evaluate and analyze code during compile-time. One such application is using `QuasiQuoter`s, which allows us to run custom parsers during compile time. Here we will use them for two purposes: to use S-expressions as Haskell expressions, and to use S-expressions as Haskell patterns, both of which can be parsed during compilation.
+
+We will create a new file which we'll call `TH.hs`, whose definition is given below.
+
+```hs
+module TH where
+
+import Control.Monad ((<=<))
+import Language.Haskell.TH.Quote (QuasiQuoter (..))
+import Language.Haskell.TH.Syntax (Q, dataToPatQ, liftData)
+
+import SExp (SExp, parseSExp)
+
+sexp :: QuasiQuoter
+sexp = QuasiQuoter
+  { quoteExp  = liftData <=< parseOrThrow
+  , quotePat  = dataToPatQ (const Nothing) <=< parseOrThrow
+  , quoteType = const $ fail "No S-expression parser for type"
+  , quoteDec  = const $ fail "No S-expression parser for declarations"
+  }
+  where
+    parseOrThrow :: String -> Q SExp
+    parseOrThrow = either fail pure . parseSExp
+```
+
+Additionally, `liftData` and `dataToPatQ` assume that our parsed `SExp` implement the `Data` type class. We will change our `SExp` module to automatically derive this datatype for us. First, enable the `DeriveDataTypeable` language extension, so GHC can automatically implement `Data` for us.
+
+You also need to import `Data` from `Data.Data`.
+
+```hs
+{-# LANGUAGE DeriveDataTypeable #-}
+
+module SExp where
+
+import Data.Data (Data)
+-- The rest of the file goes here...
+```
+
+And now change `Identifier` and `SExp` to derive `Data`:
+
+```hs
+newtype Identifier = Identifier
+  { getId :: String
+  } deriving (Data, Show)
+
+data SExp
+  = SSExp    SExp [SExp]  -- (foo 0 "hello" bar), (bar (baz 1)), (foo)
+  | SInteger Integer  -- 42
+  | SString  String  -- "hello, world"
+  | SBool    Bool  -- false, true
+  | SId      Identifier  -- foo
+  | SDouble  Double  -- 42f, 3.1415f
+  deriving (Data, Show)
+```
+
+The basic contents for the `TH` module are summarized as follows:
```

Comment:
would link the full file here or somewhere if you have it, since some mistakes with arriving to the final one can appear here, I believe

Reply:
The full file for the TH module is just above this part, it's just a pretty small file, and the complete code for the parser is linked just before the QuasiQuotation part.